File: app/src/main/java/com/vgaw/scaffolddemo/data/auth/AuthRepository.java
package com.vgaw.scaffolddemo.data.auth;

import com.vgaw.scaffolddemo.data.user.UserInfoCache;
import com.vgaw.scaffold.http.bean.HttpResult;
import com.vgaw.scaffold.http.callback.HttpResultCallback;
import com.vgaw.scaffolddemo.model.AuthInfo;

import okhttp3.ResponseBody;
import retrofit2.Call;

/**
 * @author caojin
 * @date 2018/6/6
 */
public class AuthRepository implements AuthDataSource {
    private static volatile AuthRepository sInstance = null;

    private AuthRepository() {
    }

    public static AuthRepository getInstance() {
        if (sInstance == null) {
            synchronized (AuthRepository.class) {
                if (sInstance == null) {
                    sInstance = new AuthRepository();
                }
            }
        }
        return sInstance;
    }


    @Override
    public Call<ResponseBody> login(String account, String password, HttpResultCallback<HttpResult<AuthInfo>> callback) {
        afterLogin(null);
        return null;
    }

    private void afterLogin(AuthInfo authInfo) {
        UserInfoCache.getInstance().setAuthInfo(authInfo);
    }
}

File: app/src/main/java/com/vgaw/scaffolddemo/data/auth/AuthRepository.kt
package com.vgaw.scaffolddemo.data.auth

import com.haoman.schedule.data.auth.remote.AuthRemoteDataSource
import com.vgaw.scaffold.http.callback.Result
import com.vgaw.scaffolddemo.data.HttpUtil
import com.vgaw.scaffolddemo.data.user.UserCacheDataSource

class AuthRepository {
    private constructor()

    companion object {
        private val INSTANCE by lazy { AuthRepository() }

        fun getInstance() = INSTANCE
    }

    suspend fun login(accountNo: String, authorization: String, password: String): Result<Unit> {
        try {
            val result = HttpUtil.proResponse(AuthRemoteDataSource.getInstance().login(accountNo, authorization, password)!!)
            if (result.code == Result.ResultCode.SUCCESS) {
                UserCacheDataSource.getInstance().setToken(result.data.accessToken)
                return Result.buildSuccess(Unit)
            } else {
                return Result.buildError(result.error)
            }
        } catch (e: Exception) {
            return Result.buildError(e.localizedMessage)
        }
    }
}